### PR TITLE
Minimax ai

### PR DESCRIPTION
**Pull request by maysakanoni**
We have had to restructure the design to make the AI work. At the moment we have duplicate code in the domain object and evaluate use case. This is because the `AI` depends on the `FindWinningCombination` use case and `EvaluateBoard` use case and we are still looking for a clean solution. The test need A LOT of refactoring- we will fix this. Sorry in advance for the huge commit- we just want some general feedback on the architecture/structure.

### CODE REVIEW

File: bin/game.rb [hunk 1]
```diff
@@ -14,18 +16,23 @@ def initialize
     @game_ui = UI.new(stdout: STDOUT, stdin: STDIN)
     game = Game.new(board_width: 3)
     @board_gateway = InMemoryBoardGateway.new(game.board)
+    @board_gateway_hack = InMemoryBoardGateway.new(game)
```
Comment (yusufsheiqh) [inline bin/game.rb:19]: This is very suspicious looking...

Comment (dushan-madetech) [reply bin/game.rb:19]: agree with Yusuf here

File: bin/game.rb [hunk 2]
```diff
@@ -14,18 +16,23 @@ def initialize
...
 
   def execute
     @game_ui.start
     display_current_board
-    while @evaluate_board.execute == :continue
+    while @evaluate_board.execute[:outcome] == :continue
       prompt_player_to_place_mark
       begin
-        place_players_mark
+        @active_player == 'X' ? place_players_mark : place_ai_mark
```
Comment (dushan-madetech) [inline bin/game.rb:34]: is the player always X?

File: bin/game.rb [hunk 3]
```diff
@@ -5,6 +5,8 @@
 require_relative '../lib/ui_board'
 require_relative '../lib/use_case/update_board'
 require_relative '../lib/use_case/evaluate_board'
+require_relative '../lib/use_case/find_winning_combinations'
+require_relative '../lib/use_case/ai_response_board'
 require_relative '../lib/gateway/board_gateway'
```
Comment (yusufsheiqh) [inline bin/game.rb:10]: There is no need to require `board_gateway` as it no longer exists(?).

File: bin/game.rb [hunk 4]
```diff
@@ -14,18 +16,23 @@ def initialize
...
     display_current_board
-    while @evaluate_board.execute == :continue
+    while @evaluate_board.execute[:outcome] == :continue
       prompt_player_to_place_mark
       begin
-        place_players_mark
+        @active_player == 'X' ? place_players_mark : place_ai_mark
+        # place_ai_mark
       rescue RangeError
         exception_message("\nPlease choose a valid mark\n")
```
Comment (dushan-madetech) [inline bin/game.rb:37]: I would expect a RangeError to mention something being out of range in the error message

File: bin/game.rb [hunk 5]
```diff
@@ -40,6 +47,13 @@ def execute
 
   private
 
+  def place_ai_mark
+    puts "board is: #{@board_gateway_hack.fetch_board.board}"
```
Comment (dushan-madetech) [inline bin/game.rb:51]: why not just clone the board and use that in minimax instead?

Also, fetch_board.board could be renamed -  I'd expect fetch board to return the board without using .board

**Comment by yusufsheiqh** [comment]
Did the `game.rb` file need to be committed/pushed to this PR? There are a lot of discrepancies within.

File: lib/domain/board.rb [hunk 6]
```diff
@@ -5,5 +5,42 @@ class Game
   def initialize(board_width:)
     size = board_width * board_width
     @board = Array.new(size)
+    @winning_combinations = [
```
Comment (yusufsheiqh) [inline lib/domain/board.rb:8]: Is this necessary? I assume the `find_winning_combinations` use case does this already.

https://github.com/C-gyorfi/tic-tac-toe/blob/24af5772e02f80f05b595c651c828afd55cd6aa1/lib/domain/board.rb#L8-L17

Comment (yusufsheiqh) [reply lib/domain/board.rb:8]: There is also inconsistency with file name and class name.

File: bin/game.rb (hunk 1 above)
Comment (C-gyorfi) [reply bin/game.rb:19]: When you forget to remove your hacks before pushing to github, the game/UI is incomplete... 😅

File: lib/domain/board.rb (hunk 6 above)
Comment (C-gyorfi) [reply lib/domain/board.rb:8]: Fixed

**Pull request merged**